exercises/practice/collatz-conjecture/.approaches/worker-wrapper/content.md

```diff
@@ -0,0 +1,198 @@
+
+```haskell
+_ =  length "abcde"
+  == 1 + length "bcde"
+  == 1 + (1 + length "cde")
+  == 1 + (1 + (1 + length "de"))
+  == 1 + (1 + (1 + (1 + length "e")))
+  == 1 + (1 + (1 + (1 + (1 + length ""))))
+  -- this computation is as large as the list was long!
+  == 1 + (1 + (1 + (1 + (1 + 0))))
+  == 1 + (1 + (1 + (1 + 1)))
+  == 1 + (1 + (1 + 2))
+  == 1 + (1 + 3)
+  == 1 + 4
+  == 5
+```
+
+It would be more efficient to keep track of a running total number of encountered elements.
+To do this we can use a variant of `length` with an additional parameter:
+
+```haskell
+length' :: Int -> [a] -> Int
+length' count []     = count
+length' count (_:xs) = length' (1 + count) xs
+```
+
+Here the parameter `count` is used to keep track of the number of previously encountered elements.
+It might evaluate as follows.
+
+```haskell
+_ =  length' 0 "abcde"
+  == length' 1 "bcde"
+  == length' 2 "cde"
+  == length' 3 "de"
+  == length' 4 "e"
+  == length' 5 ""
+  == 5
+```
+
+The computation stays small at every step &ndash; much more efficient!
+
+To obtain an efficient function `length` with the familiar `[a] -> Int` type &ndash; instead of the more involved `Int -> [a] -> Int` &ndash; we can 'wrap' `length'` in a function that provides it with initial arguments:
+
+```haskell
+length :: [a] -> Int
+length xs = length' 0 xs
+  where
+    length' count []     = count
+    length' count (_:xs) = length' (1 + count) xs
+```
+
+Here `length` is known as the _wrapper_, and `length'` as the _worker_.
+Hence, this is an example of the _worker&ndash;wrapper_ construct.
+
+The worker is commonly called `go`, `loop`, or `f'` when the wrapper is called `f`.
+
+The worker&ndash;wrapper pattern is more general than demonstrated here.
+For a few more more examples, see the [relevant wiki article][wiki-worker-wrapper], and for yet more examples and an explanation of the general pattern see the paper «[The worker/wrapper transformation][paper-worker-wrapper]» (pdf).
+
+
+## Bang patterns
+
+The above implementation of `length'` _might_ evaluate as efficiently as illustrated.
+However, it also might not!
+The definition of `length'` allows evaluation to proceed as follows as well.
+
+```haskell
+_ =  length' 0 "abcde"
+  == length' (1 + 0) "bcde"
+  == length' (1 + (1 + 0)) "cde"
+  == length' (1 + (1 + (1 + 0))) "de"
+  == length' (1 + (1 + (1 + (1 + 0)))) "e"
+  == length' (1 + (1 + (1 + (1 + (1 + 0))))) ""
+  == 1 + (1 + (1 + (1 + (1 + 0))))
+  == 1 + (1 + (1 + (1 + 1)))
+  == 1 + (1 + (1 + 2))
+  == 1 + (1 + 3)
+  == 1 + 4
+  == 5
+```
+
+Still a computation as large as the list was long is built.
+The situation is not quite as bad as before though.
+
+In the case of the naive implementation of `length`, the built up expression has the form
+
+```haskell
+_ = 1 + (1 + (⋯(1 + length xs)⋯))
+```
+
+This expression cannot be simplified before `length xs` has been evaluated.
```

Comment:
~~~suggestion
In the case of the naive implementation of `length`, the built up expression has the form

```haskell
_ = 1 + (1 + (⋯(1 + length …)⋯))
```

This expression cannot be simplified before `length …` has been evaluated.
~~~